[NVIDIA GPU] Use memcpy for intra-node all-to-all

terryysun opened the pull request:
The communications of all-to-all rely on NCCL even when it is intra-node. By leveraging memcpy for intra-node communications, all-to-all can have better performance while reducing SM consumption (right now consumed by NCCL).

terryysun commented:
some screenshot showing the performance difference between NCCL call and memcpy for an example all-to-all.
<img width="896" alt="Screenshot 2024-07-19 at 2 24 44 PM" src="https://github.com/user-attachments/assets/1ac2b485-45be-46e8-9396-0bbd7c35930f">
<img width="925" alt="Screenshot 2024-07-19 at 2 24 05 PM" src="https://github.com/user-attachments/assets/45f3cfef-cbbe-45a6-9c60-b65a149ee405">

frgossen reviewed (changes requested):
Very nice! 

You are adding this for all-gather, all-reduce, all-to-all, collective-broadcast, and collective-permute but I only see a test case for all-to-all. 
Can we add test cases for the remaining collectives? would also suggest to split this up into one PR per collective to keep reviews simpler and in case we have to roll back something.

terryysun commented:
> Very nice!
> 
> You are adding this for all-gather, all-reduce, all-to-all, collective-broadcast, and collective-permute but I only see a test case for all-to-all. Can we add test cases for the remaining collectives? would also suggest to split this up into one PR per collective to keep reviews simpler and in case we have to roll back something.

Thanks! Right now it's only added for all-to-all, the api changes for the other collectives are just to avoid interim intricacies -- we do plan to add similar support for them in the near future, but the flag is not consumed in this PR.

frgossen commented on xla/service/gpu/runtime/nccl_collective_thunk.h:
```diff
@@ -71,6 +71,21 @@ struct NcclCollectiveConfig {
 };
 
+template <typename T>
+absl::StatusOr<const int64_t> GetCurrentId(
```
Rather than templating, would it make sense to impl this base on the NcclCollectiveConfig?

terryysun replied:
after reorganizing the code turns out we don't need this util, removed.

frgossen commented on xla/service/gpu/runtime/nccl_all_to_all_thunk.h:
```diff
@@ -37,9 +40,55 @@ struct NcclAllToAllConfig {
+
+    absl::Status PutRecvPtr(int64_t send_id, int64_t recv_id, void* ptr) {
+      if (!IsInitialized(send_id, recv_id)) {
+        return absl::InternalError(absl::StrCat("Send-receive pair ", send_id, ", ", recv_id,
```
nit: recv?

terryysun replied:
updated to receive.

frgossen commented on xla/service/gpu/runtime/nccl_all_to_all_thunk.cc:
```diff
@@ -99,22 +98,37 @@ absl::Status NcclAllToAllStartThunk::RunNcclCollective(
+                               comm_wrapper.comm_handle, current_id, use_memcpy, recv_ptr_map_);
 }
 
 absl::Status RunAllToAll(NcclApi* nccl_api, bool has_split_dimension,
```
This is implemented in two modes which do not share much code. Can we outline that into two functions and dispatch here?

terryysun replied:
broke to two functions.

frgossen reviewed (changes requested):
Thanks for clarifying. All makes sense to me, just a few minot comments.

frgossen commented:
I see you pushed changes and requested a review. Can you reply to the comments and explain how the changes address them? Ty.

terryysun commented:
> I see you pushed changes and requested a review. Can you reply to the comments and explain how the changes address them? Ty.

hey @frgossen sorry for the delayed reply, we were verifying the changes and fixed multiple issues we saw when running realistic models. I've updated the code and replied to the comments accordingly. Could you take another look? Thanks!

frgossen reviewed (approved):
Thanks!

dimitar-asenov commented:
Hi @terryysun. This PR causes threading issues with the test `tests:collective_ops_e2e_test_2gpu`. It seems like there is a race condition:

```
[ RUN      ] CollectiveOpsTestE2E.AsyncAllToAllMemCpy

assert.h assertion failed at absl/container/internal/raw_hash_set.h:3976 in void absl::container_internal::raw_hash_set<absl::container_internal::NodeHashMapPolicy<long, absl::flat_hash_map<long, unsigned long>>, absl::hash_internal::Hash<long>, std::equal_to<long>, std::allocator<std::pair<const long, absl::flat_hash_map<long, unsigned long>>>>::AssertNotDebugCapacity() const [Policy = absl::container_internal::NodeHashMapPolicy<long, absl::flat_hash_map<long, unsigned long>>, Hash = absl::hash_internal::Hash<long>, Eq = std::equal_to<long>, Alloc = std::allocator<std::pair<const long, absl::flat_hash_map<long, unsigned long>>>]: capacity() != InvalidCapacity::kReentrance &&
"Reentrant container access during element construction/destruction " "is not allowed."

assert.h assertion failed at absl/container/internal/raw_hash_set.h:3976 in void absl::container_internal::raw_hash_set<absl::container_internal::NodeHashMapPolicy<long, absl::flat_hash_map<long, unsigned long>>, absl::hash_internal::Hash<long>, std::equal_to<long>, std::allocator<std::pair<const long, absl::flat_hash_map<long, unsigned long>>>>::AssertNotDebugCapacity() const [Policy = absl::container_internal::NodeHashMapPolicy<long, absl::flat_hash_map<long, unsigned long>>, Hash = absl::hash_internal::Hash<long>, Eq = std::equal_to<long>, Alloc = std::allocator<std::pair<const long, absl::flat_hash_map<long, unsigned long>>>]: capacity() != InvalidCapacity::kReentrance &&
"Reentrant container access during element construction/destruction " "is not allowed."

*** Check failure stack trace: ***
    @     0x7fb643bddcc3  __assert_fail
    @     0x7fbe4963ac12  absl::container_internal::raw_hash_set<>::AssertNotDebugCapacity()
    @     0x7fbe496474fc  absl::container_internal::raw_hash_set<>::AssertOnFind<>()
    @     0x7fbe496473b3  absl/container/internal/raw_hash_set.h:3443 absl::container_internal::raw_hash_set<>::find<>()
    @     0x7fbe49647143  absl/container/internal/raw_hash_set.h:3457 absl::container_internal::raw_hash_set<>::find<>()
    @     0x7fbe4964c6e5  absl/container/internal/raw_hash_set.h:3536 absl::container_internal::raw_hash_set<>::FindElement::operator()<>()
    @     0x7fbe4964c665  absl/container/internal/container_memory.h:153 absl::container_internal::memory_internal::DecomposePairImpl<>()
    @     0x7fbe4964c5b8  absl/container/internal/container_memory.h:220 absl::container_internal::DecomposePair<>()
    @     0x7fbe4964c543  absl/container/node_hash_map.h:704 absl::container_internal::NodeHashMapPolicy<>::apply<>()
    @     0x7fbe4964bb33  absl/container/internal/hash_policy_traits.h:134 absl::container_internal::hash_policy_traits<>::apply<>()
    @     0x7fbe4964a0a5  absl/container/internal/raw_hash_set.h:4056 absl::container_internal::raw_hash_set<>::emplace_at<>()
    @     0x7fbe49649ee1  absl/container/internal/raw_hash_map.h:214 absl::container_internal::raw_hash_map<>::try_emplace_impl<>()
    @     0x7fbe49649e1f  absl/container/internal/raw_hash_map.h:133 absl::container_internal::raw_hash_map<>::try_emplace<>()
    @     0x7fbe496309c7  absl/container/internal/raw_hash_map.h:184 absl::container_internal::raw_hash_map<>::operator[]<>()
    @     0x7fbe496277ed  xla/service/gpu/runtime/nccl_all_to_all_thunk.cc:121 xla::gpu::NcclAllToAllStartThunk::Initialize()
    @     0x7fbc5e7c4b8a  xla/service/gpu/runtime/sequential_thunk.cc:69 xla::gpu::SequentialThunk::Initialize()
    @     0x7fbc655dca9a  xla/service/gpu/gpu_executable.cc:464 xla::gpu::(anonymous namespace)::ExecuteThunks()
    @     0x7fbc655da74d  xla/service/gpu/gpu_executable.cc:1011 xla::gpu::GpuExecutable::ExecuteAsyncOnStreamImpl()
    @     0x7fbc655dab3f  xla/service/gpu/gpu_executable.cc:798 xla::gpu::GpuExecutable::ExecuteAsyncOnStream()
```

The test above, when run in the 2 gpu configuration fails ~ 80% of the times. I tried to fix it by adding a mutex that protects `send_pointer_maps_` and `receive_pointer_maps_`, but that didn't quite work. It definitely eliminated the problem above, but then the test started deadlocking in about 1 out of 1000 runs. I'm not sure if the deadlock was caused by my attempts at a fix, or if it's already in the code and my fix simply exposed it.

Could you please take a look?

dimitar-asenov commented:
I disabled the flaky test in https://github.com/openxla/xla/pull/17641. Please reenable as part of the fix.

loislo commented:
From time to time a test  xla/tests/collective_ops_e2e_test.cc:486 xla::(anonymous namespace)::CollectiveOpsTestE2E_AsyncAllToAllMemCpy_Test::TestBody() is failing due to a thread gets stuck.

This is the stacktrace for this thread.

E1002 05:16:35.729531    5068 rendezvous.cc:43] Stack for thread tf_replicas/5068:
   @      clock_gettime
   @      third_party_gpus_cuda_compat_cuda_compat_data_o_tmpdir_filewrapper_s0_start
   @      third_party_gpus_cuda_compat_cuda_compat_data_o_tmpdir_filewrapper_s0_start
   @      third_party_gpus_cuda_compat_cuda_compat_data_o_tmpdir_filewrapper_s0_start
   @      third_party_gpus_cuda_compat_cuda_compat_data_o_tmpdir_filewrapper_s0_start
   @      third_party_gpus_cuda_compat_cuda_compat_data_o_tmpdir_filewrapper_s0_start
   @      xla/stream_executor/cuda/cuda_driver.cc:1358 stream_executor::gpu::GpuDriver::SynchronizeStream()
   @      xla/stream_executor/cuda/cuda_executor.cc:528 stream_executor::gpu::CudaExecutor::BlockHostUntilDone()
   @      xla/stream_executor/stream_common.cc:148 stream_executor::StreamCommon::BlockHostUntilDone()
   @      xla/service/gpu/runtime/nccl_all_to_all_thunk.cc:313 xla::gpu::RunMemCpyAllToAll()
   @      xla/service/gpu/runtime/nccl_all_to_all_thunk.cc:189 xla::gpu::NcclAllToAllStartThunk::RunNcclCollective()
   @      xla/service/gpu/runtime/nccl_collective_thunk.cc:447 xla::gpu::NcclCollectiveThunk::ExecuteOnStream()
   @      xla/service/gpu/runtime/sequential_thunk.cc:81 xla::gpu::SequentialThunk::ExecuteOnStream()
   @      xla/service/gpu/gpu_executable.cc:481 xla::gpu::(anonymous namespace)::ExecuteThunks()
   @      xla/service/gpu/gpu_executable.cc:1011 xla::gpu::GpuExecutable::ExecuteAsyncOnStreamImpl()
   @      xla/service/gpu/gpu_executable.cc:798 xla::gpu::GpuExecutable::ExecuteAsyncOnStream()
   @      xla/service/executable.cc:84 xla::Executable::ExecuteOnStream()
   @      xla/service/hlo_runner.cc:607 xla::HloRunner::ExecuteReplicated()::$_0::operator()()::{lambda()#2}::operator()()
   @      Thread::ThreadBody()


We are waiting this thread from the other one that triggers the watchdog code.
   @      xla/service/rendezvous.cc:83 xla::internal::AwaitAndLogIfStuck()
   @      ./xla/service/rendezvous.h:307 xla::RendezvousSingle<>()
   @      ./xla/service/rendezvous.h:336 xla::RendezvousSingle<>()
   @      ./xla/service/rendezvous.h:361 xla::RendezvousSingle<>()
   @      xla/service/gpu/runtime/nccl_collective_thunk.cc:486 xla::gpu::NcclCollectiveThunk::ExecuteOnStream()
   @      xla/service/gpu/runtime/sequential_thunk.cc:81 xla::gpu::SequentialThunk::ExecuteOnStream()
   @      xla/service/gpu/gpu_executable.cc:481 xla::gpu::(anonymous namespace)::ExecuteThunks()
   @      xla/service/gpu/gpu_executable.cc:1011 xla::gpu::GpuExecutable::ExecuteAsyncOnStreamImpl()
   @      xla/service/gpu/gpu_executable.cc:798 xla::gpu::GpuExecutable::ExecuteAsyncOnStream()
   @      xla/service/executable.cc:84 xla::Executable::ExecuteOnStream()
   @      xla/service/hlo_runner.cc:607 xla::HloRunner::ExecuteReplicated()::$_0::operator()()::{lambda()#2}::operator()()

loislo commented:
@terryysun it is failing on the machine with p100 gpus

terryysun commented:
> @terryysun it is failing on the machine with p100 gpus

thanks! will file a followup pr to make it work on P100 and make the flag default, right now it's an optional flag so should not affect anything. it's [verified](https://github.com/openxla/xla/pull/17636#issuecomment-2387831208) on H100s. working on reproducing the P100 failure on our side
